README


This archive contains the distribution HTML-Element-AbsoluteXPath,
version 0.02:

  Add absolute XPath support to HTML::Element

This software is copyright (c) 2013 by HyeonSeung Kim.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




CHANGES
==================================================
Changes from 2012-03-15 00:00:00 +0000 to present.
==================================================

-----------------------------------------
version 0.02 at 2013-03-15 06:14:42 +0000
-----------------------------------------

  Change: b1a7a49b104589285f71d130d1d5fe84e1212479
  Author: HyunSeung Kim <sng2nara@hanmail.net>
  Date : 2013-03-15 15:13:46 +0000

    fix error - comparing element 

  Change: 9d36dc2bcd9d6f58287ba54f07cf283f95523704
  Author: HyunSeung Kim <sng2nara@hanmail.net>
  Date : 2013-03-15 14:40:45 +0000

    fix bug 

  Change: dd45bc3d765e7b7d73dd3a5ef4810ea23cea1af5
  Author: HyunSeung Kim <sng2nara@hanmail.net>
  Date : 2013-03-15 14:38:01 +0000

    up to 0.01

    add code to remove whitespaces of attributes. 

  Change: c635618f7144885dea15a303e92400cb5e3a16cc
  Author: HyunSeung Kim <sng2nara@hanmail.net>
  Date : 2013-03-15 14:32:33 +0000

    remove necessary codes 

  Change: 819b7ee7af5119a33e0803ab23d2a9f398086fb0
  Author: HyunSeung Kim <sng2nara@hanmail.net>
  Date : 2013-03-13 17:23:45 +0000

    fix error on test 

  Change: 26f99d22e8858b20f96d472215f7fbe8227c336e
  Author: HyunSeung Kim <sng2nara@hanmail.net>
  Date : 2013-03-13 17:21:25 +0000

    fix errors on examples

    update requirements right. 

  Change: 4ee06b6341c81fed489fcdbec1f58b24c909cbe1
  Author: HyunSeung Kim <sng2nara@hanmail.net>
  Date : 2013-03-13 15:20:40 +0000

    update README.md 

  Change: 94bf6effcdc92555945534c0cb668141f179ceec
  Author: HyunSeung Kim <sng2nara@hanmail.net>
  Date : 2013-03-13 15:19:06 +0000

    add use HTML::Element in module.

    modify dist.ini to fill Changes. 

  Change: ad15808064338cd9d7bb3faa8f4bf0b06cca8e26
  Author: HyunSeung Kim <sng2nara@hanmail.net>
  Date : 2013-03-13 14:53:00 +0000

    remove backslash in POD. 

  Change: 70b994298700a78e31e4266db00a4258c0ebb7ce
  Author: HyunSeung Kim <sng2nara@hanmail.net>
  Date : 2013-03-13 14:47:49 +0000

    add [1] to root path. 

  Change: 36859729ff53f061c655b584586b5003f1053966
  Author: HyunSeung Kim <sng2nara@hanmail.net>
  Date : 2013-03-13 14:30:19 +0000

    fix README 

  Change: 936969ba7f816a5aab7a29cf473e32ed0bee0e60
  Author: HyunSeung Kim <sng2nara@hanmail.net>
  Date : 2013-03-13 14:28:17 +0000

    fix error on example. 

  Change: 66b7ae13c16b7af877b00dbb2a91ab91f69b3d27
  Author: HyunSeung Kim <sng2nara@hanmail.net>
  Date : 2013-03-13 14:24:57 +0000

    modify dist.ini 

  Change: 157167b5676ef33383a26e2b0186e1c16b4faf1f
  Author: HyunSeung Kim <sng2nara@hanmail.net>
  Date : 2013-03-13 14:21:56 +0000

    initial 

================
End of releases.
================


lib/HTML/Element/AbsoluteXPath.pm
use strict;
use warnings;
package HTML::Element::AbsoluteXPath;

# ABSTRACT: Add absolute XPath support to HTML::Element

our $VERSION = '0.02'; # VERSION

use HTML::Element 5.03;

sub HTML::Element::abs_xpath{
    my $self = shift;
    my @hint = @_;
   
    push(@hint,'_tag') unless grep{$_ eq '_tag'}@hint;

    my $raddr = $self->root()->address;
    my @lin;
    my $ee = $self; 
    while( $ee->address ne $raddr ){
        my $p = $ee->parent();
        my %filters;
        foreach (sort @hint){
            my $v = $ee->attr($_);
            next unless $v;
            $filters{$_} = $v;
        }   
        my @sib = $p->look_down(%filters);

        @sib = grep{$_->depth() == $ee->depth()}@sib;
        my $idx = 1;
        foreach (@sib){
            last if $_->address eq $ee->address;
            $idx++;
        }   

        my @attrs;
        foreach (keys %filters){
            next if $_ eq '_tag';
            my $v = $filters{$_};
            my $pat = "\@$_='$v'";
            push(@attrs,$pat);
        }   

        my $attr = ''; 
        $attr = "[".join(' and ',@attrs)."]" if @attrs;

        push(@lin, $ee->tag() . $attr . "[$idx]" );
        $ee = $p; 
    }   
    push(@lin, $ee->tag().'[1]' );
    return '/'.join('/',reverse @lin);
}




1;

__END__

=pod

=head1 NAME

HTML::Element::AbsoluteXPath - Add absolute XPath support to HTML::Element

=head1 VERSION

version 0.02

=head1 DESCRIPTION

HTML::Element::AbsoluteXPath adds ABSOLUTE XPath support to HTML::Element by adding 'abs_xpath' method to HTML::Element package.

It generates smarter XPaths with HINTS which are attributes name of HTML element, like 'class','id','width','name' and etc.

=head1 SYSNOPSIS

    use HTML::Element;
    use HTML::Element::AbsoluteXPath;

    ...

    print $elem->abs_xpath;
    # output like '/html[1]/body[1]/div[1]'

more

    use 5.010;
    use HTML::TreeBuilder->new;
    use HTML::Element::AbsoluteXPath;
    
    my $root = HTML::TreeBuilder->new;
    my $html = <<END;
    <html>
        <body>
            <div id="test" class="testclass"></div>
            <div class="testclass"></div>
            <div>
                <div class="innerclass"></div>
                <div></div>
            </div>
        </body>
    </html>
    END
    $root->parse($html);
    $root->eof();

    # get abs xpath of root
    say $root->abs_xpath; # '/html[1]' 
    
    my @found = $root->find_by_tag_name('div');
    
    # get abs xpath
    say $found[0]->abs_xpath; # '/html[1]/body[1]/div[1]' 
    
    # get abs xpath with 'id' hint.
    say $found[0]->abs_xpath('id'); #, "/html[1]/body[1]/div[@id='test'][1]"
    
    # get abs xpath with 'id' and 'class' hints.
    say $found[0]->abs_xpath('id','class'); # "/html[1]/body[1]/div[@id='test' and @class='testclass'][1]"
    
    # get abs xpath hints for elem has just 'class' attr.
    say $found[1]->abs_xpath('id','class'); # "/html[1]/body[1]/div[@class='testclass'][2]"
    
    # get abs xpath with hints for elem has no attrs
    say $found[2]->abs_xpath('id','class'); # "/html[1]/body[1]/div[3]"
    
    # get abs xpath overwrapped one
    say $found[2]->content->[0]->abs_xpath('id','class'); # "/html[1]/body[1]/div[3]/div[@class='innerclass'][1]"
    
    # get abs xpath overwrapped sibling
    say $found[2]->content->[1]->abs_xpath('id','class'); # "/html[1]/body[1]/div[3]/div[2]"

=head1 METHODS

=head3 abs_xpath( [ $hint, ... ] )

Returns an absolute xpath of current HTML::Element object.

Hints are optional. Internally, '_tag' is added to hints.

With no hints, it generates an xpath with only ordinary expressions like "/html[1]/body[1]/div[1]", "/html[1]/body[1]/a[13]".

If you set 'class' as hint, it generates a smarter xpath like "/html[1]/body[1]/div[@class='SOME'][1]","/html[1]/body[1]/div[2]".

Hints work for being matched attribute names which have values.

=head1 SEE ALSO

=over

=item *

L<HTML::TreeBuilder>

=item *

L<HTML::TreeBuilder::XPath>

=item *

L<HTML::Element>

=back

=head1 SOURCE

L<https://github.com/sng2c/HTML-Element-AbsoluteXPath>

=head1 AUTHOR

HyeonSeung Kim <sng2nara@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by HyeonSeung Kim.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Add absolute XPath support to HTML::Element",
  "AUTHOR" => "HyeonSeung Kim <sng2nara\@gmail.com>",
  "BUILD_REQUIRES" => {
    "HTML::TreeBuilder" => "5.03",
    "Test::More" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "HTML-Element-AbsoluteXPath",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "HTML::Element::AbsoluteXPath",
  "PREREQ_PM" => {
    "HTML::Element" => "5.03"
  },
  "VERSION" => "0.02",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



